src/lib.rs

```diff
@@ -370,48 +606,123 @@ impl TestHarness {
-            "Tested node is not a validator",
-        );
-        let height = self.state().height();
-        let last_hash = *self.state().last_hash();
-        let round = Round::first();
+        let height = self.height().next();
+        let last_hash = self.last_block_hash();
 
-        let handler = &mut self.handler;
-        let (block_hash, patch) = handler.create_block(validator_id, height, tx_hashes);
-        handler.state.add_block(
-            block_hash,
-            patch,
-            tx_hashes.to_vec(),
-            validator_id,
-        );
+        self.update_configuration();
 
-        let propose = self.network.us().create_propose(
-            height,
-            &last_hash,
-            tx_hashes,
-        );
-        let precommits: Vec<_> = self.network
+        let (block_hash, patch) = {
+            let validator_id = self.leader().validator_id().expect(
+                "Tested node is not a validator",
+            );
+            let transactions = self.mempool();
+            self.blockchain.create_patch(
+                validator_id,
+                height,
+                tx_hashes,
+                &transactions,
+            )
+        };
+
+        // Remove txs from mempool
+        {
+            let mut transactions = self.mempool.write().expect(
+                "Cannot modify transactions in mempool",
+            );
+            for hash in tx_hashes {
+                transactions.remove(hash);
+            }
+        }
+
+        let propose = self.leader().create_propose(height, &last_hash, tx_hashes);
+        let precommits: Vec<_> = self.network()
             .validators()
             .iter()
             .map(|v| v.create_precommit(&propose, &block_hash))
             .collect();
-        handler.commit(block_hash, precommits.iter(), Some(round));
+
+        self.blockchain
+            .commit(&patch, block_hash, precommits.iter())
+            .unwrap();
+
+        self.poll_events();
+    }
+
+    /// Update test network configuration if such an update has been scheduled
+    /// with `commit_configuration_change`.
+    fn update_configuration(&mut self) {
+        use ConfigurationProposalState::*;
+
+        let height = self.height().next().next();
+        if let Some(cfg_proposal) = self.cfg_proposal.take() {
+            match cfg_proposal {
+                Uncommitted(cfg_proposal) => {
+                    // Commit configuration proposal
+                    let stored = cfg_proposal.stored_configuration().clone();
+                    let mut fork = self.blockchain.fork();
+                    CoreSchema::new(&mut fork).commit_configuration(stored);
+                    let changes = fork.into_patch();
+                    self.blockchain.merge(changes).unwrap();
+                    self.cfg_proposal = Some(Committed(cfg_proposal));
+                }
+                Committed(ref cfg_proposal) if cfg_proposal.actual_from() == height => {
```

Comment:
Nitpick: I wonder whether it would make sense to take the condition into the body of the arm:

```rust
Committed(cfg_proposal) => {
    if cfg_proposal.actual_from() == height {
        // ...
    } else {
        // ...
    }
}
```

It could prevent cloning network params below. I'm not quite sure which way is more Rustic, of course 🤔

Reply:
Fixed.